pyomo/core/base/block.py

```diff
@@ -1857,22 +1880,27 @@ def _getitem_when_not_present(self, idx):
             if data is not None:
                 data = data.get(idx, None)
             if data is not None:
+                # Note that for scalar Blocks, this will override the
+                # entry for _BlockConstruction.data[id(self)], as _block
+                # is self.
                 _BlockConstruction.data[id(_block)] = data
         else:
             data = None
 
         try:
-            obj = apply_indexed_rule(
-                self, self._rule, _block, idx, self._options)
+            obj = self._rule(_block, idx)
+            # If the user returns a block, transfer over everything
+            # they defined into the empty one we created.  We do
+            # this inside the try block so that any abstract
+            # components declared by the rule have the opportunity
+            # to be initialized with data from
+            # _BlockConstruction.data as they are transferred over.
+            if obj is not _block and isinstance(obj, _BlockData):
+                _block.transfer_attributes_from(obj)
```

Comment:
Just for my understanding, does this work for classes that inherit from blocks? I think so...

Reply:
Yes.  This should work for derived blocks.  That said, if the derived block class has non-component attributes that hold state information that shouldn't be moved over, then they need to define a `_Block_reserved_words` class attribute that extends the set defined on `Block._Block_reserved_words`.